### PR TITLE
feat(events): API Destinations

### LINKED ISSUE #13607: (aws-events-targets): Support new EventBridge API Destinations feature
Add construct to support event targets using API Destinations

### Use Case

Looking to integrate with external parties as direct EventBridge targets
https://aws.amazon.com/blogs/compute/using-api-destinations-with-amazon-eventbridge/

### Proposed Solution
Follow established patterns as add this as a new target type

* [X] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by msimpsonnz**
Adds `Connection` and `ApiDestination` constructs in Events, and an `ApiDestination` target in Events Targets. Can be used to do arbitrary HTTP calls.

Fixes #13729 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by msimpsonnz** [comment]
@rix0rrr It would be great to get a first review on this make sure I’m in the right path

File: packages/@aws-cdk/aws-events/lib/api-destination.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  /**
+   * The ARN of the Api Destination created.
+   * @attribute
+   */
+  public readonly apiDestinationArn: string;
+
+  constructor(scope: Construct, id: string, props: ApiDestinationProps) {
+    super(scope, id, {
+      physicalName: props.apiDestinationName || Lazy.string({
```
Comment (DaWyz) [inline packages/@aws-cdk/aws-events/lib/api-destination.ts:87]: Would this work as well ? 

```ts
physicalName: props.apiDestinationName || Lazy.string({ produce: () => Names.uniqueId(this) })
```

The `generateUniqueName()` method seems to limit the name to 21 characters max but I cannot see requirements for this in the CFN doc.

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+   * @param scope Parent construct
+   * @param id Construct ID
+   * @param connectionArn ARN of imported connection
+   */
+  public static fromEventBusArn(scope: Construct, id: string, connectionArn: string, connectionSecretArn: string): IConnection {
+    const parts = Stack.of(scope).parseArn(connectionArn);
+
+    return new ImportedConnection(scope, id, {
+      connectionArn: connectionArn,
+      connectionName: parts.resourceName || '',
```
Comment (DaWyz) [inline packages/@aws-cdk/aws-events/lib/connection.ts:269]: I think we should consider the `arn` invalid if `parts.resourceName` does not exist.

```ts
if (!parts.resourceName) {
      throw new Error('invalid events connection ARN');
}
```

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+export interface ClientParameters {
+  /**
+   * The client secret associated with the client ID to use for OAuth authorization for the connection.
+   */
+  readonly clientID: string;
+
+  /**
+   * The client ID to use for OAuth authorization for the connection.
+   */
+  readonly clientSecret: string;
```
Comment (DaWyz) [inline packages/@aws-cdk/aws-events/lib/connection.ts:60]: Should this be a `cdk.SecretValue` instead of a string ?

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+}
+
+/**
+ * Contains the Basic authorization parameters to use for the connection.
+ */
+export interface BasicAuthParameters {
+  /**
+   * The password associated with the user name to use for Basic authorization.
+   */
+  readonly password: string;
```
Comment (DaWyz) [inline packages/@aws-cdk/aws-events/lib/connection.ts:40]: Should this be a `cdk.SecretValue` instead of a string ?

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+export interface ApiKeyAuthParameters {
+  /**
+   * The name of the API key to use for authorization.
+   */
+  readonly apiKeyName: string;
+
+  /**
+   * The value for the API key to use for authorization.
+   */
+  readonly apiKeyValue: string;
```
Comment (DaWyz) [inline packages/@aws-cdk/aws-events/lib/connection.ts:30]: Should this be a `cdk.SecretValue` instead of a string ?

Have a look at the [GitHub Action for CodePipeline](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-codepipeline-actions/lib/github/source-action.ts) for an example.

Same applies for other secrets below. Wondering if we should set the `apiKeyName`/`username`/`clientId` as secretValues as well ?

**Review by DaWyz — COMMENTED** [review]
Nice job @msimpsonnz . I added a few comments on your PR. I'm no `aws-cdk` maintainer so feel free to act on them if they make sense to you or ask confirmation to @rix0rrr if needed.

File: packages/@aws-cdk/aws-events-targets/lib/api-destination.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+   * @default none
+   */
+  readonly queryStringParameters?: { [key: string]: string };
+}
+
+/**
+ * Use an API Destination rule target.
+ * @experimental
+ */
+export class ApiDestination implements events.IRuleTarget {
```
Comment (SvenBayer) [inline packages/@aws-cdk/aws-events-targets/lib/api-destination.ts:42]: I would suggest to rename this to EventBridgeApiDestination and pass ApiDestination as parameter, same as the pattern of CloudWatchLogGroup and LogGroup.

File: packages/@aws-cdk/aws-events-targets/lib/api-destination.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      pathParameterValues: this.props.pathParameterValues,
+      queryStringParameters: this.props.queryStringParameters,
+    };
+
+    return {
+      arn: this.connection.connectionArn,
+      input: this.props.event,
+      targetResource: this.connection,
+      httpParameters,
+    };
```
Comment (SvenBayer) [inline packages/@aws-cdk/aws-events-targets/lib/api-destination.ts:64]: Since we add the role as required property, we should assign it here and return it in this bind method.

File: packages/@aws-cdk/aws-events-targets/lib/api-destination.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+   */
+  public bind(_rule: events.IRule, _id?: string): events.RuleTargetConfig {
+    const httpParameters: events.CfnRule.HttpParametersProperty = {
+      headerParameters: this.props.headerParameters,
+      pathParameterValues: this.props.pathParameterValues,
+      queryStringParameters: this.props.queryStringParameters,
+    };
+
+    return {
+      arn: this.connection.connectionArn,
```
Comment (SvenBayer) [inline packages/@aws-cdk/aws-events-targets/lib/api-destination.ts:60]: Since we add the TargetBaseProps, we should return ...bindBaseTargetConfig(this.props), as in CloudWatchLogGroup

File: packages/@aws-cdk/aws-events-targets/lib/api-destination.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  readonly queryStringParameters?: { [key: string]: string };
+}
+
+/**
+ * Use an API Destination rule target.
+ * @experimental
+ */
+export class ApiDestination implements events.IRuleTarget {
+  constructor(
+    private readonly connection: events.Connection,
```
Comment (SvenBayer) [inline packages/@aws-cdk/aws-events-targets/lib/api-destination.ts:44]: Here, we should use the ApiDestination that actually holds the Connection.

**Review by SvenBayer — CHANGES_REQUESTED** [review]
@msimpsonnz, your pull request is pretty cool. I tried your code but could not get it working to add an Api Destination to my EventBridge Rule. However, I realised that the **CloudWatchLogGroup** is an IRuleTarget that wraps **LogGroup** that contains all the information. I would suggest to use here the same pattern and to wrap ApiDestination in **aws-events** with ApiDestination in **aws-event-targets**.

After my experiment I got it working. You can see my suggested changes in my comments. I hope they are helpful.

File: packages/@aws-cdk/aws-events-targets/lib/api-destination.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import * as events from '@aws-cdk/aws-events';
+
+/**
+ * Customize the EventBridge Api Destinations Target
+ * @experimental
+ */
+export interface ApiDestinationProps {
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-events-targets/lib/api-destination.ts:7]: A role makes sense but if possible should be optional and automatically generated for the user if we can.

File: packages/@aws-cdk/aws-events-targets/lib/api-destination.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+   */
+  readonly event?: events.RuleTargetInput;
+
+  /**
+   * The headers that need to be sent as part of request invoking the API Gateway REST API
+   * or EventBridge Api destinations.
+   *
+   * @default none
+   */
+  readonly headerParameters?: { [key: string]: string };
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/lib/api-destination.ts:21]: Why are these here? Aren't they API gateway specific?

Doesn't it make more sense to separate this out into API Gateway destinations and "other" destinations?

File: packages/@aws-cdk/aws-events-targets/README.md [hunk 12]
```diff
@@ -28,6 +28,7 @@ Currently supported are:
 * [Log an event into a LogGroup](#log-an-event-into-a-loggroup)
 * Put a record to a Kinesis Data Firehose stream
 * Put an event on an EventBridge bus
+* Send an event to EventBridge API Destination
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-events-targets/README.md:31]: Yes I really personally need this to also even understand what the feature is and does.

[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-events-targets/README.md:31]: Is this for arbitrary HTTP calls?

File: packages/@aws-cdk/aws-events-targets/test/api-destination/api-destination.test.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
+import { expect, haveResource } from '@aws-cdk/assert';
+import * as events from '@aws-cdk/aws-events';
+import { Stack } from '@aws-cdk/core';
+import * as targets from '../../lib';
+
+test('use aws batch job as an eventrule target', () => {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/test/api-destination/api-destination.test.ts:6]: What's the AWS Batch part here?

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+    if (props.authorizationType === AuthorizationType.BASIC) {
+      if (props.authParameters?.basicAuthParameters !== undefined) {
+        authParameters = {
+          BasicAuthParameters: {
+            Password: props.authParameters?.basicAuthParameters?.password,
+            Username: props.authParameters?.basicAuthParameters?.username,
+          },
+        };
+      } else {
+        throw new Error('You must supply the relevant authParameters for this authorizationType');
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:333]: You can be more specific than this! You know the auth type so you know what field they need to be passing.

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 15]
```diff
@@ -0,0 +1,423 @@
...
+            ClientParameters: {
+              ClientID: props.authParameters?.oAuthParameters?.clientParameters.clientID,
+              ClientSecret: props.authParameters?.oAuthParameters?.clientParameters.clientSecret,
+            },
+            HttpMethod: props.authParameters?.oAuthParameters?.httpMethod,
+            //OAuthHttpParameters: props.authParameters?.oAuthParameters?.oAuthHttpParameters,
+          },
+        };
+      } else {
+        throw new Error('You must supply the relevant authParameters for this authorizationType');
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:351]: Also here

**Review by rix0rrr — CHANGES_REQUESTED** [review]
Thanks for the submission!

There are a couple of small things that I'd like to see addressed (most of them already pointed out by @DaWyz), and I guess I'm not too fond of the name "API Destination" because it seems it could have been a lot more descriptive (how about `HttpDestination`?) but I guess we're bound to whatever EventBridge chose as their terminology...

**Comment by msimpsonnz** [comment]
> Thanks for the submission!
> 
> There are a couple of small things that I'd like to see addressed (most of them already pointed out by @DaWyz), and I guess I'm not too fond of the name "API Destination" because it seems it could have been a lot more descriptive (how about `HttpDestination`?) but I guess we're bound to whatever EventBridge chose as their terminology...

Thanks  @rix0rrr, I'm happy to rename but this is what EventBridge had chosen, I wasn't sure if changing this might affect functionality down the road.

**Comment by msimpsonnz** [comment]
Thanks @DaWyz and @SvenBayer 
This was work in progress, I wanted to check I was on the right path with the different auth methods
I've updated and this now has a working deployment with a role created if you don't supply one

File: packages/@aws-cdk/aws-events-targets/README.md [hunk 16]
```diff
@@ -229,3 +230,44 @@ rule.addTarget(
...
+    },
+  },
+  description: 'ConnectionDescription',
+  connectionName: 'testConnection',
+});
+  
+const destination = new events.ApiDestination(this, 'Destination', {
+  apiDestinationName: 'apiDestinationName',
+  connection: connection,
+  invocationEndpoint: 'https://example.com,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/README.md:263]:
```suggestion
  invocationEndpoint: 'https://example.com',
```

File: packages/@aws-cdk/aws-events/lib/connection.ts (hunk 5 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-events/lib/connection.ts:30]: Let me rephrase:

Password and keys DEFINITELY need to be typed as `cdk.SecretValue`.

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 17]
```diff
@@ -0,0 +1,449 @@
...
+   * Contains the OAuth authorization parameters to use for the connection.
+   * @default - none
+   */
+  readonly oAuthHttpParameters?: HttpParameters;
+}
+
+/**
+ * Contains the authorization parameters to use to authorize with the endpoint.
+ */
+export interface AuthParameters {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:135]: How is this structure supposed to be used? I assume at most one of the auth types needs to be supplied?

In this case, I think I'd prefer you model this as an [enum-like class](https://github.com/aws/aws-cdk/blob/master/docs/DESIGN_GUIDELINES.md#enums):

```ts
export class AuthParameters {
  public static basicAuth(username: string, password: SecretValue): AuthParameters {
    return new AuthParameters({
      authorizationType: AuthorizationType.BASIC,
      basicAuthParameters: { username, password },
    });
  }

  public static ...() {
  }

  constructor(public readonly authParametersJson: any) {
  }
}
```

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+export interface BaseConnectionProps {
+  /**
+   * The name of the connection.
+   *
+   * @default - none
+   */
+  readonly authorizationType: AuthorizationType;
+
+  /**
+   * The name of the connection.
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:173]:
```suggestion
   * Authentication method for this Connection
```

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 19]
```diff
@@ -0,0 +1,449 @@
...
+   *
+   * @default - none
+   */
+  readonly authParameters: AuthParameters;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:177]:
```suggestion
  readonly authParameters?: AuthParameters;
```

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 20]
```diff
@@ -0,0 +1,449 @@
...
+   * The ARN for the secret created for the connection.
+   * @attribute
+   */
+  readonly connectionSecretArn: string;
+}
+
+/**
+ * The event connection properties
+ */
+export interface ConnectionProps extends BaseConnectionProps {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:228]: Duplicate definition

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 21]
```diff
@@ -0,0 +1,449 @@
...
+   * @default - none
+   */
+  readonly connectionName: string;
+}
+
+
+/**
+ * The event connection properties
+ */
+export interface ConnectionProps extends BaseConnectionProps {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:198]: Why even have this empty interface?

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 22]
```diff
@@ -0,0 +1,449 @@
...
+   * The name of the connection.
+   *
+   * @default - none
+   */
+  readonly description: string;
+
+  /**
+   * The name of the connection.
+   *
+   * @default - none
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:189]: Why is this `@default - none` ? Either a name is required, in which case there is no `@default` at all, or the name should be generated in which case it's not `none`.

If the name is required, can we make it non-required?

File: packages/@aws-cdk/aws-events/lib/connection.ts [hunk 23]
```diff
@@ -0,0 +1,449 @@
...
+   */
+  readonly authParameters: AuthParameters;
+
+  /**
+   * The name of the connection.
+   *
+   * @default - none
+   */
+  readonly description: string;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/connection.ts:184]:
```suggestion
  readonly description?: string;
```

File: packages/@aws-cdk/aws-events/lib/api-destination.ts [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+  /**
+   * A description for the API destination.
+   *
+   * @default - none
+   */
+  readonly description?: string;
+  /**
+   * The method to use for the request to the HTTP invocation endpoint.
+   */
+  readonly httpMethod: HttpMethod;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/api-destination.ts:46]: Does it make sense to make this optional and default it to `POST` ?

File: packages/@aws-cdk/aws-events/lib/api-destination.ts [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+  /**
+   * The URL to the HTTP invocation endpoint for the API destination..
+   */
+  readonly invocationEndpoint: string;
+  /**
+   * The maximum number of requests per second to send to the HTTP invocation endpoint.
+   * @default - None
+   */
+  readonly invocationRateLimit?: Duration;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/api-destination.ts:55]: The description says the unit of this value is Hz (Hertz, N per second), so why is its declared type `Duration` ?

Right now I could pass:

```ts
new ApiDestination(..., {
  invocationRule: Duration.hours(1),
})
```

What does that mean?

File: packages/@aws-cdk/aws-events/lib/api-destination.ts [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+/**
+ * The event Api Destination properties
+ */
+export interface ApiDestinationProps {
+  /**
+   * The ARN of the connection to use for the API destination
+   */
+  readonly connection: IConnection;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/api-destination.ts:36]:
```suggestion
  readonly connection?: IConnection;
```

The docs say that `connection` is optional. That probably corresponds to an anonymous request?

I wonder if we can make the API even easier by making it look like:

```ts
export interface ApiDestinationProps {
  /** 
   * The connection to use for this API Destination
   *
   * @default - Create a new connection from `authParameters` if those are given,
   * otherwise make an anonymous call.
   */  
  readonly connection?: IConnection;

  /**
   * Authentication to use for this API Destination (HTTP call)
   *
   * If `connection` is not supplied but `authParameters` are, implicitly
   * create a new `Connection` resource based on the specified auth parameters.
   *
   * This is convenient if you only have one or a few API destinations. If you want
   * to share authentication between multiple API destinations, create and pass
   * a `Connection` resource explicitly.
   *
   * @default - Use the existing connection supplied via `connection`, 
   * otherwise make an anonymous call.
   */
  readonly authParameters?: AuthParameters;
}
```

That way you don't even have to explicitly create a `Connection` object in the simple case.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-events/lib/api-destination.ts (hunk 25 above)
Comment (msimpsonnz) [reply packages/@aws-cdk/aws-events/lib/api-destination.ts]: Fixed

File: packages/@aws-cdk/aws-events/lib/connection.ts (hunk 22 above)
Comment (msimpsonnz) [reply packages/@aws-cdk/aws-events/lib/connection.ts:207]: Resolved

File: packages/@aws-cdk/aws-events/lib/connection.ts (hunk 17 above)
Comment (msimpsonnz) [reply packages/@aws-cdk/aws-events/lib/connection.ts:135]: Thanks for the feedback @rix0rrr 
I'm a little confused on how to proceed, is there another similar example in the code base of a enum-like class?
I if use `authParametersJson` as `any` then I end up needing to parse out the `authorisationType` when it comes to building the object
```
let connection = new CfnConnection(this, 'Connection', {
authorizationType: props.authParameters.authParametersJson.authorizationType,
authParameters: props.authParameters.authParametersJson,
description: props.description,
name: this.physicalName,
});
```
I also need to figure out how to build the header and query strings for each connection

**Comment by nhuray** [comment]
Thanks @msimpsonnz for that PR, it's really more convenient than dealing with the Cfn constructs ! 

Do we have an ETA for releasing it ?

File: packages/@aws-cdk/aws-events/lib/api-destination.ts (hunk 26 above)
Comment (nhuray) [reply packages/@aws-cdk/aws-events/lib/api-destination.ts:36]: @rix0rrr I didn't find the documentation stating `connection` is optional. But in the AWS UI it's clearly required.
Screen Shot 2021-08-04 at 4 18 06 PM

File: packages/@aws-cdk/aws-events-targets/test/api-destination/api-destination.test.ts [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+    },
+    description: 'ConnectionDescription',
+    connectionName: 'testConnection',
+  });
+
+  const destination = new events.ApiDestination(stack, 'Destination', {
+    apiDestinationName: 'apiDestinationName',
+    connection: connection,
+    invocationEndpoint: 'https://endpoint.com',
+    invocationRateLimit: Duration.seconds(10),
```
Comment (nhuray) [inline packages/@aws-cdk/aws-events-targets/test/api-destination/api-destination.test.ts:27]: @msimpsonnz  `invocationRateLimit` should be the `The maximum number of requests per second to send to the HTTP invocation endpoint.`. 

Why in that test you are setting `Duration.seconds(10)` instead of `10` ?

File: packages/@aws-cdk/aws-events/test/test.connection.ts [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+        },
+      },
+      Name: 'testConnection',
+      Description: 'ConnectionDescription',
+    }));
+
+    test.done();
+  },
+
+  // 'create connection with InvocationHttpParameters'(test: Test) {
```
Comment (nhuray) [inline packages/@aws-cdk/aws-events/test/test.connection.ts:34]: @msimpsonnz Does that test fails ?

File: packages/@aws-cdk/aws-events-targets/lib/api-destination.ts (hunk 11 above)
Comment (nikp) [reply packages/@aws-cdk/aws-events-targets/lib/api-destination.ts:27]: HttpParameters (headers/body/querystrings) on EventBridge targets are re-used for both ApiGateway and ApiDestinations targets

File: packages/@aws-cdk/aws-events/lib/connection.ts (hunk 5 above)
Comment (nikp) [reply packages/@aws-cdk/aws-events/lib/connection.ts]: EventBridge consideres api key names, usernames, and client Ids as non-secret, and their value counterparts secrets

On the other hand, headers, query strings, and body parameters can be classified as secret/non-secret by the customer.

**Comment by nikp** [comment]
Hey @msimpsonnz are you still interested in finishing this? I can provide a round of code review...

**Comment by msimpsonnz** [comment]
@nikp yeah I am keen to continue, just wanted a review on path forward as felt there is a bit to do with Cfn generators

**Review by rix0rrr — DISMISSED** [review]

**Review by rix0rrr — APPROVED** [review]

**Review by rix0rrr — APPROVED** [review]
